File: src/Tuple.cs
using System;

namespace RayTracerChallenge
{
    abstract class Tuple
    {
        const float EPSILON = 0.00001f;

        public float x;
        public float y;
        public float z;
        public byte w;

        public Tuple(float x, float y, float z, byte w)
        {
            this.x = x; 
            this.y = y; 
            this.z = z;
            this.w = w;
        }

        public static Tuple operator +(Tuple t1, Tuple t2)
        {
            if (t1 is Color && t2 is Color)
                return new Color(t1.x + t2.x, t1.y + t2.y, t1.z + t2.z);

            if (t1 is Point && t2 is Point)
                throw new InvalidOperationException("Both tuples are points and cannot be added.");

            if (t1 is Point || t2 is Point)
                return new Point(t1.x + t2.x, t1.y + t2.y, t1.z + t2.z);

            else
                return new Vector(t1.x + t2.x, t1.y + t2.y, t1.z + t2.z);
        }

        public static Tuple operator -(Tuple t1, Tuple t2)
        {
            if (t1 is Color && t2 is Color)
                return new Color(t1.x - t2.x, t1.y - t2.y, t1.z - t2.z);

            if (t1 is Point && t2 is Point)
                return new Vector(t1.x - t2.x, t1.y - t2.y, t1.z - t2.z);

            if (t1 is Point || t2 is Point)
                return new Point(t1.x - t2.x, t1.y - t2.y, t1.z - t2.z);

            else
                return new Vector(t1.x - t2.x, t1.y - t2.y, t1.z - t2.z);
        }

        public static Tuple operator -(Tuple t)
        {
            if (t is Vector)
                return new Vector(-t.x, -t.y, -t.z);

            else
                return new Point(-t.x, -t.y, -t.z);
        }

        public static Tuple operator *(Tuple t, float c)
        {
            if (t is Vector)
                return new Vector(t.x * c, t.y * c, t.z * c);

            else if (t is Point)
                return new Point(t.x * c, t.y * c, t.z * c);

            else
                return new Color(t.x * c, t.y * c, t.z * c);
        }

        public static Tuple operator /(Tuple t, float c)
        {
            if (t is Vector)
                return new Vector(t.x / c, t.y / c, t.z / c);

            else
                return new Point(t.x / c, t.y / c, t.z / c);
        }

        public void Print()
        {
            if (w == 0)
                Console.WriteLine("({0}, {1}, {2} | VECTOR)", x, y, z);

            else if (w == 1)
                Console.WriteLine("({0}, {1}, {2} | POINT)", x, y, z);

            else
                Console.WriteLine("({0}, {1}, {2} | COLOR)", x, y, z);
        }
    }
}

File: src/Tuple.cpp
#include "Tuple.hpp"

#include <iostream>
#include <sstream>
#include <iomanip>
#include <cmath>

Tuple::Tuple() {}

Tuple::Tuple(float x, float y, float z, int w) : x(x), y(y), z(z), w(w)
{
    if (w != 0 && w != 1) throw std::invalid_argument("Attempted to create an invalid tuple type!");
}

Tuple Tuple::Vector(float x, float y, float z)
{
    return Tuple(x, y, z, 0);
}

Tuple Tuple::Point(float x, float y, float z)
{
    return Tuple(x, y, z, 1);
}

float Tuple::magnitude()
{
    return sqrt(pow(x, 2) + pow(y, 2) + pow(z, 2));
}

Tuple Tuple::normalize()
{
    return Tuple::Vector(x / magnitude(), y / magnitude(), z / magnitude());
}

float Tuple::dot(const Tuple& a, const Tuple& b)
{
    return (a.x * b.x) + (a.y * b.y) + (a.z * b.z);
}

Tuple Tuple::cross(const Tuple& a, const Tuple& b)
{
    return Tuple::Vector(a.y * b.z - a.z * b.y, a.z * b.x - a.x * b.z, a.x * b.y - a.y * b.x);
}

bool Tuple::operator==(const Tuple& other) const
{
    return std::fabs(x - other.x) < EPSILON &&
           std::fabs(y - other.y) < EPSILON &&
           std::fabs(z - other.z) < EPSILON &&
           w == other.w;
}

Tuple Tuple::operator+(const Tuple& other) const
{
    if (is_point() && other.is_point()) throw std::invalid_argument("Attempted to add two points!");
    return Tuple(x + other.x, y + other.y, z + other.z, w + other.w);
}

Tuple Tuple::operator-(const Tuple& other) const
{
    if (is_vector() && other.is_point()) throw std::invalid_argument("Attempted to subtract point from vector!");
    return Tuple(x - other.x, y - other.y, z - other.z, w - other.w);
}

Tuple Tuple::operator*(const float scalar) const
{
    return Tuple(x * scalar, y * scalar, z * scalar, w);
}

Tuple Tuple::operator/(const float scalar) const
{
    if (scalar == 0) throw std::invalid_argument("Attempted to divide by zero!");
    return Tuple(x / scalar, y / scalar, z / scalar, w);
}

Tuple Tuple::operator-() const
{
    return Tuple(-x, -y, -z, w);
}

bool Tuple::is_vector() const
{
    return w == 0;
}

bool Tuple::is_point() const
{
    return w == 1;
}

void Tuple::print() const
{
    std::ostringstream stream;
    stream << std::fixed << std::setprecision(2);
    stream << "{ " << x << ", " << y << ", " << z << " | " << (w == 0 ? "VECTOR" : w == 1 ? "POINT" : "UNDEFINED") << " }";
    std::cout << stream.str() << std::endl;
}